contracts/allocators/OnsenAllocatorV2.sol

```diff
@@ -0,0 +1,174 @@
+
+    address immutable masterChef; // Onsen contract
+
+    // address immutable treasury;
+    address public constant treasury = 0x9A315BdF513367C0377FB36545857d12e85813Ef; // Olympus Treasury
+
+    /* ========== CONSTRUCTOR ========== */
+
+    constructor(
+        address _chef,
+        address _sushi,
+        address _xSushi,
+        AllocatorInitData memory data
+    ) BaseAllocator(data) {
+        require(_chef != address(0));
+        masterChef = _chef;
+        require(_sushi != address(0));
+        sushi = _sushi;
+        require(_xSushi != address(0));
+        xSushi = _xSushi;
+
+        // approve for safety, yes toke is being instantly sent to treasury and that is fine
+        // but to be absolutely safe this one approve won't hurt
+        IERC20(sushi).approve(address(extender), type(uint256).max);
+        IERC20(xSushi).approve(address(extender), type(uint256).max);
+    }
+
+    /**
+     1- Deposit LP token into onsen, if deposit is succesfull we should get in return sushi tokens
+     2- Stake the Sushi rewards tokens returned from deposit LP tokens and as rewards form onsen LP stake, we should get in return xSushi tokens
+     3- Calculate gains/loss
+     */
+    function _update(uint256 id) internal override returns (uint128 gain, uint128 loss) {
+        // reads
+        uint256 index = tokenIds[id];
+        IERC20 LPtoken = _tokens[index];
+        uint256 balance = LPtoken.balanceOf(address(this));
+
+        // interactions
+        //Deposit LP token into onsen, if deposit succesfull this address should have in return sushi tokens
+        if (balance > 0) {
+            IERC20(LPtoken).approve(masterChef, balance);
+            IMasterChef(masterChef).deposit(id, balance); // deposit into Onsen
+        }
+
+        // Stake the sushi tokens
+        if (IERC20(sushi).balanceOf(address(this)) > 0) {
+            enterSushiBar(true); // manage rewards
+        }
+
+        //Calculate gains/loss
+        // Retrieve current balance for pool and address
+        UserInfo memory currentUserInfo = IMasterChef(masterChef).userInfo(id, address(this));
+
+        uint256 last = extender.getAllocatorAllocated(id) + extender.getAllocatorPerformance(id).gain;
+
+        if (currentUserInfo.amount >= last) {
+            gain = uint128(currentUserInfo.amount - last);
+        } else {
+            loss = uint128(last - currentUserInfo.amount);
+        }
+    }
+
+    /**
+        Asuming the index of the amounts [] can be use as parameter for the mapping tokenIds
+        Example:
+        amounts => [20,50,150]
+        So the values that will be used to retrieve the id used in onsen will be 0, 1 , 2
+    */
+    function deallocate(uint256[] memory amounts) public override onlyGuardian {
+        for (uint256 i; i < amounts.length; i++) {
+            uint256 amount = amounts[i];
+            if (amount > 0) {
+                IMasterChef(masterChef).withdraw(tokenIds[i], amount); // withdraw from Onsen
```

Comment:
Is this how you exit or are you calling `enterSushiBar(false)` ?

Reply:
IMasterChef(masterChef).withdraw that's the function to unstake the LP tokens, 
 I am adding the functionality to unstake sushi tokens.

Reply:
Again here the proper ids need to be used since you cannot repurpose tokenIds.

Reply:
This function was changed for stakeSushi.